lib/improver/utilities/cube_metadata.py

```diff
@@ -690,3 +692,39 @@ def extract_diagnostic_name(cube_name):
     # 'probability_of_' is a 15-character string
     diagnostic_name = cube_name[15:relative_to_threshold_index]
     return diagnostic_name
+
+
+def generate_hash(data_in):
+    """
+    Generate a hash from the data_in that can be used to uniquely identify
+    equivalent data_in.
+
+    Args:
+        data_in (any):
+            The data from which a hash is to be generated. This can be of any
+            type that can be pickled.
+    Returns:
+        hash (str):
+            A hexidecimal hash representing the data.
+    """
+    hashable_type = pickle.dumps(data_in)
+    hash_result = hashlib.md5(hashable_type).hexdigest()
+    return hash_result
+
+
+def create_coordinate_hash(cube):
+    """
+    Generate a hash based on the input cube's x and y coordinates. This
+    acts as a unique identifier for the grid which can be used to allow two
+    grids to be compared.
+
+    Args:
+        cube (iris.cube.Cube):
+            The cube from which x and y coordinates will be used to
+            generate a hash.
+    Returns:
+        coordinate_hash (string):
+            A hash created using the x and y coordinates of the input cube.
+    """
+    hashable_data = [cube.coord(axis='x'), cube.coord(axis='y')]
```

Comment:
I wonder if it's worth removing coordinate bounds before matching here?  I'm wondering if there are situations where we might have coordinates that don't exactly match in metadata (eg one has bounds and one doesn't), but the points and units are identical, and therefore we would want them to "match" as grids.

Reply:
I'm not sure we want to cater for those situations. What I'm after here is some pretty rigorous enforcement of the grid definitions that were used to generate the spot neighbours being exactly the same as those from which we are extracting diagnostic data. What you propose would most likely be fine, I cannot foresee a situation in in which it would be invalid/not work, but I am not famed for my prescience...

I can do it if you think the added flexibility is important, but I'm hoping IMPROVER is dealing with some pretty darn uniform grids.

Reply:
OK.  I tend to agree with you, this shouldn't be strictly necessary.  I'm just not sure what we have at the moment will necessarily conform to what we expect (the plugins do a lot of different things with metadata).  But we can fix that at source as and when.